### DISCUSSION TITLE
Przeniesienie aplikacji do Azure'a

### DISCUSSION THREAD

**Discussion by tomaszkyc** [AzureCommunityPL/group#84] [Pomoc] [answered]
Cześć,
Mam pytanie dotyczące umieszenia mojej aplikacji w Azure. Stack: 
Front: Angular 9,
Backend: Java 11, Spring
DB: Sql Server 2019 Linux
Obecnie na localu mam to skonteneryzowane (Docker compose), natomiast chciałbym to przenieść jaknajtaniej i najszybciej na Azure. Obecnie myślę o takim schemacie jak poniżej:
Frontend - AppService
Backend - AppService
DB - Azure SQL Database
Połączenie między Frontend i Backend to Virtual Network, to samo pomiędzy Backend oraz DB - tak, aby nie można było przejść z sieci Frontendowej do sieci, gdzie jest DB. 
Bardzo proszę o pomoc 🙂

**Comment by kaluzaaa** [comment]
Gadasz do backnedu poprzez jakieś proxy z nginx/npm/node w tym kontenerze z frontend?

**Reply by tomaszkyc** [reply]
Tak mam proxy postawione na npm i przy uruchamianiu aplikacji w Angularze dodaję parametr wskazujący na plik proxy ([przykładowy config jak w tym artykule](https://medium.com/better-programming/setup-a-proxy-for-api-calls-for-your-angular-cli-app-6566c02a8c4d)).

Daj znać, czy Ci o to chodziło :)

**Comment by kaluzaaa** [comment, marked as answer]
Uprość to:
- frontend na świat w jakimś blob statycznym (albo kontener z nginx) i dopuść ruch tylko z adresacji front door
- backend na świat wystaw i dopuść ruch tylko z adresacji front door
- przed tym azure front door, który robi routing na apkę a jak jest /api to na backend
- backend wpinasz do sieci razem z bazą https://docs.microsoft.com/en-us/azure/app-service/web-sites-integrate-with-vnet i https://docs.microsoft.com/en-us/azure/azure-sql/database/private-endpoint-overview

**Reply by tomaszkyc** [reply]
Dzięki wielkie - spróbuję to tak zrobić i dam znać :)

**Reply by smereczynski** [reply]
@kaluzaaa mądrego to aż miło poczytać 👍

**Reply by ptrstpp950** [reply]
To ja zaproponuje jeszcze jedno uproszczenie całej sprawy. Wywal blob, o którym @kaluzaaa pisał. Zamiast tego na app plan wgraj docker compose z front + backend (jest to w wersji preview). Dzięki temu albo:
- masz tak jak Łukasz napisał, tylko bez bloba (możesz /api udostępnić na app plan)
- albo robisz tak jak na stacji deweloperskiej (czyli z plikiem proxy)

Dlaczego tak? Testowałem z blob i jest on upierdliwy we wdrażaniu.

**Reply by mgrabarz** [reply]
@ptrstpp950 zalecałbym ostrożność z Docker Compose w AppService. 
Ta funkcja jest w preview of kilku lat, nie jest rozwijana, mocno ograniczona i już dwa lata temu słyszałem, że NIGDY nie opuści preview.

Zalecane metody hostowania stron statycznych to na serio z Bloba lub w planie Windows https://azure.microsoft.com/pl-pl/services/app-service/static/ lub ostatecznie z planu Linux np z Apache. Ten compose to raczek.
Co do API, powinno być bezpiecznie wystawione na publicznym IP, mieć włączone VNet Integration i sięgać do bazy wewnątrz sieci po Private Linku.

**Reply by mgrabarz** [reply]
A jednak nie o Docker Compose chodziło, a o Kubernetesa. Nie zmienia to faktu, że Compose w wykonaniu AppService ma wiele niedoróbek.
https://github.com/Azure-Samples/multicontainerwordpress/issues/4

**Comment by piotrgwiazda** [comment]
Hej

Nie czuję sensu robienia vNet dla PaaS. W tym vNet nic nie będzie,
AppService jest poza vNet na infrze współdzielonej. Ewentualnie AppService
Environment ale to kalkulator pęka przy obliczniach.

AppService ma wbudowane (wliczone w cenę) obsługę SSL i akceptację ruchu
publicznego, monitoring, telemetrię, staging i inne takie. Jak nie chcemy z
tego korzystać, tylko zdeployować kontener, to jest Container Instance.

Czuję sens izolacji się vNetami jak mamy izolację własnej infry od poziomu
VM. Np Kubernetes (da się go włączyć nawet na 1 maszynie serii B). Wtedy
private linki itp.

Kiedy i tak idziemy w PaaS to vNet będzie służył jako przejściówka pomiędzy
dwoma PaaS'ami.


niedz., 27 wrz 2020, 22:34 użytkownik Marek Grabarz <
notifications@github.com> napisał:

**Reply by kamil-mrzyglod** [reply]
Ale tutaj chyba rozminąłeś się z konceptem o którym pisał @kaluzaaa - nie chodziło o robienie wielkiego VNet dla PaaS tylko wykorzystanie Private Link do zabezpieczenia bazy. Dzięki czemu już nie możesz np. tak radośnie wynosić danych z bazy nawet jak ktoś da radę wynieść do niej credentiale bo nie będzie w tej samej sieci a publiczny dostęp jest zamknięty.

**Comment by piotrgwiazda** [comment]
Przeczytałem jeszcze raz swoją odpowiedź, trochę nie brzmi tak jak chciałem.

Faktycznie, to prawda, że ograniczenie dostępu do bazy dla podsieci, lub
Private Link, a potem użycie tej integracyjnej podsieci dla App Service
podnosi bezpieczeństwo bazy.
Bez tego musimy włączyć dostęp dla usług Azure, a to oznacza wszystkie
usługi, także z innych subskrybcji.

Chodzi mi o to, że "pójście w vNety" zwykle bardzo utrudnia życie, i w
modelu PaaS warto rozważyć, czy na pewno trzeba.

W moim świecie klienci raczej odrzucają PaaS na starcie i tak z różnych
powodów. Ten hybrydowy model PaaS+vNet integracyjny daje ciekawe
możliwości, ale dokłada złożoność do tego problemu.



pon., 28 wrz 2020, 20:10 użytkownik Kamil Mrzygłód <notifications@github.com>
napisał:

**Reply by kamil-mrzyglod** [reply]
I to prawda :) To jest referencyjna architektura, która bardzo mocno kryje tyłek, ale czy z definicji powinna być używana - tutaj pewnie można podyskutować.

**Reply by kaluzaaa** [reply]
@kamil-mrzyglod moim zdaniem zawsze gdzie nie jest to pet project do nauki :-) Jak sobie popatrzysz taka architektura to "dołożenie starań".

**Comment by tomaszkyc** [comment]
Udało mi się wdrożyć podane 3 usługi, gdzie front, backend oraz db są wystawione na świat. Front i backend muszą pozostać wystawione, ponieważ daję możliwość integracji z moim backendem innym systemom/aplikacjom. 

Pytanie moje brzmi: Czy aby schować bazę przed światem mogę zrobićtakie coś: 
1. Stworzyć VNet.
2. Dodać backend do tego VNetu
3. Dodać bazę do tego VNetu?


Jeśli tak - proszę o ew. wskazówki i linki do dokumentacji. Z góry dziękuję :)

**Reply by kamil-mrzyglod** [reply]
No ale tutaj już @kaluzaaa Ci wrzucał linki w jednym ze swoich postów w tym wątku.

**Comment by piotrgwiazda** [comment]
Mam na wierzchu
https://docs.microsoft.com/pl-pl/azure/app-service/web-sites-integrate-with-vnet

wt., 29 wrz 2020, 19:01 użytkownik Kamil Mrzygłód <notifications@github.com>
napisał:

**Comment by tomaszkyc** [comment]
Bardzo dziękuję wszystkim za pomoc :)
